src/examples/mnist/MNIST_SP.cpp

```diff
@@ -0,0 +1,240 @@
+    }
+    if(magic_number != 0x00000803) {
+        cerr << "ERROR: MNIST data is compressed or corrupt" << endl;
+        exit(1);
+    }
+    NTA_ASSERT(n_rows == 28);
+    NTA_ASSERT(n_cols == 28);
+    UInt img_size = n_rows * n_cols;
+    vector<UInt*> retval;
+    for(int i = 0; i < number_of_images; ++i) {
+        auto data_raw = new unsigned char[img_size];
+        file.read( (char*) data_raw, img_size);
+        // Copy the data into an array of UInt's
+        auto data = new UInt[img_size];
+        // auto data = new UInt[2 * img_size];
+        // Apply a threshold to the image, yielding a B & W image.
+        for(UInt pixel = 0; pixel < img_size; pixel++) {
+            data[pixel] = data_raw[pixel] >= 128 ? 1 : 0;
+            // data[2 * pixel] = data_raw[pixel] >= 128 ? 1 : 0;
+            // data[2 * pixel + 1] = 1 - data[2 * pixel];
+        }
+        retval.push_back(data);
+        delete[] data_raw;
+    }
+    return retval;
+}
+
+
+int main(int argc, char **argv) {
+  UInt verbosity = 1;
+  auto train_dataset_iterations = 1u;
+  int opt;
+  while ( (opt = getopt(argc, argv, "tv")) != -1 ) {  // for each option...
+    switch ( opt ) {
+      case 't':
+          train_dataset_iterations += 1;
+        break;
+      case 'v':
+          verbosity = 1;
+        break;
+      case '?':
+          cerr << "Unknown option: '" << char(optopt) << "'!" << endl;
+        break;
+    }
+  }
+
+  SDR input({28, 28, 1});
+  SpatialPoolerExtended sp(
+    /* inputDimensions */           input.dimensions,
+    /* columnDimensions */          {10, 10, 120},
+    /* potentialRadius */           4,
+    /* potentialPct */              .9,
+    /* wrapAround */                true,
+    /* localAreaDensity */          .015,
+    /* stimulusThreshold */         28,
+    /* synPermInactiveDec */        .00928,
+    /* synPermActiveInc */          .032,
+    /* synPermConnected */          .422,
+    /* dutyCyclePeriod */           1402,
+    /* minPctOverlapDutyCycles */   0.,
+    /* seed */                      Random()(),
+    /* spVerbosity */               verbosity);
+
+  SDR columns({sp.getNumColumns()});
+  SDR_Metrics columnStats(columns, 1402);
+
+  SDRClassifier clsr(
+    /* steps */         {0},
+    /* alpha */         .001,
+    /* actValueAlpha */ .3,
+                        verbosity);
+
+  // Train
+  auto train_images = read_mnist_images("./mnist_data/train-images-idx3-ubyte");
+  auto train_labels = read_mnist_labels("./mnist_data/train-labels-idx1-ubyte");
+  if(verbosity)
+    cout << "Training for " << (train_dataset_iterations * train_labels.size())
+         << " cycles ..." << endl;
+  for(auto i = 0u; i < train_dataset_iterations; i++) {
+    // Shuffle the training data.
+    vector<UInt> index( train_labels.size() );
+    for(auto s = 0u; s < train_labels.size(); s++)
+        index[s] = s;
+    Random().shuffle( index.begin(), index.end() );
```

Comment:
reproducibility, use a seeded instance of rnd

Reply:
I'd rather add random seed CLI argument, and default it to zero?